src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,374 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/async_data_uploader.h"
+
+#include "archival/logger.h"
+#include "archival/types.h"
+#include "base/vlog.h"
+#include "model/record.h"
+#include "model/record_batch_types.h"
+#include "model/timeout_clock.h"
+#include "model/timestamp.h"
+#include "resource_mgmt/io_priority.h"
+#include "ssx/future-util.h"
+#include "storage/disk_log_impl.h"
+#include "storage/offset_to_filepos.h"
+#include "storage/parser.h"
+#include "storage/segment_appender_utils.h"
+#include "storage/segment_index.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/core/when_all.hh>
```

Comment:
nit: there are some unused imports here and in the accompanying header, I tried removing them and building to check if the IDE was misreporting but seems to build fine, maybe they can be removed?

The diff of all removed imports:

```
diff --git a/src/v/archival/async_data_uploader.cc b/src/v/archival/async_data_uploader.cc
index 07aab18949..17105ef319 100644
--- a/src/v/archival/async_data_uploader.cc
+++ b/src/v/archival/async_data_uploader.cc
@@ -14,26 +14,20 @@
 #include "archival/types.h"
 #include "base/vlog.h"
 #include "model/record.h"
-#include "model/record_batch_types.h"
 #include "model/timeout_clock.h"
-#include "model/timestamp.h"
 #include "resource_mgmt/io_priority.h"
 #include "ssx/future-util.h"
 #include "storage/disk_log_impl.h"
 #include "storage/offset_to_filepos.h"
-#include "storage/parser.h"
 #include "storage/segment_appender_utils.h"
-#include "storage/segment_index.h"
 #include "utils/retry_chain_node.h"

 #include <seastar/core/abort_source.hh>
 #include <seastar/core/loop.hh>
 #include <seastar/core/shared_ptr.hh>
-#include <seastar/core/when_all.hh>

 #include <exception>
 #include <stdexcept>
-#include <system_error>

 namespace archival {

diff --git a/src/v/archival/async_data_uploader.h b/src/v/archival/async_data_uploader.h
index dff70333d4..05386c543f 100644
--- a/src/v/archival/async_data_uploader.h
+++ b/src/v/archival/async_data_uploader.h
@@ -17,17 +17,11 @@
 #include "utils/retry_chain_node.h"

 #include <seastar/core/abort_source.hh>
-#include <seastar/core/rwlock.hh>
 #include <seastar/core/scheduling.hh>
 #include <seastar/core/shared_ptr.hh>

 #include <boost/iterator/counting_iterator.hpp>
 #include <boost/iterator/transform_iterator.hpp>
-#include <boost/outcome.hpp>
-#include <boost/outcome/outcome.hpp>
-#include <boost/outcome/policy/all_narrow.hpp>
-
-#include <exception>
 #include <utility>

 namespace archival {
```